File: app/src/main/java/com/ndmsystems/coala/layers/arq/states/LoggableState.java
package com.ndmsystems.coala.layers.arq.states;

import com.ndmsystems.coala.helpers.TimeHelper;
import com.ndmsystems.coala.message.CoAPMessage;
import com.ndmsystems.infrastructure.logging.LogHelper;

/**
 * Created by bas on 25.08.17.
 */

public abstract class LoggableState implements CoAPMessage.ResendHandler  {
    private final Long startTime;
    private Integer diffTime = null;
    private Integer numberOfMessages;
    private Integer numberOfResend;
    LoggableState() {
        startTime = TimeHelper.getTimeForMeasurementInMilliseconds();
        numberOfMessages = 0;
        numberOfResend = 0;
    }

    void onTransferCompleted() {
        LogHelper.d("onTransferCompleted");
        if (diffTime == null) diffTime = (int) (TimeHelper.getTimeForMeasurementInMilliseconds() - startTime);
    }

    public abstract int getDataSize();

    public abstract byte[] getToken();

    public long getSpeed() {
        if (diffTime == null) return -1;
        return (long) (((double)getDataSize()) / (diffTime / 1000.0));
    };

    public Double getPercentOfLoss() {
        if (numberOfMessages == 0) return null;
        return ((double) numberOfResend * 100) / (numberOfResend + numberOfMessages);
    }

    public Integer getDiffTime() {
        return diffTime;
    }

    public abstract boolean isIncoming();


    public Integer getNumberOfResend() {
        return numberOfResend;
    }

    @Override
    public void onResend() {
        numberOfResend++;
    }

    public void incrementNumberOfMessage() {
        numberOfMessages++;
    }
}

File: app/src/main/java/com/ndmsystems/coala/layers/arq/states/LoggableState.kt
package com.ndmsystems.coala.layers.arq.states

import com.ndmsystems.coala.helpers.MonotonicClock
import com.ndmsystems.coala.helpers.logging.LogHelper.d
import com.ndmsystems.coala.message.CoAPMessage.ResendHandler

abstract class LoggableState internal constructor(
    /** Seam for tests: transfer timings are measured against this, not the system clock. */
    private val clock: MonotonicClock = MonotonicClock.SYSTEM
) : ResendHandler {
    private val startTime: Long = clock.nowMillis()
    var diffTime: Int? = null
        private set
    private var numberOfMessages: Int = 0
    var numberOfResend: Int
        private set

    init {
        numberOfResend = 0
    }

    fun onTransferCompleted() {
        d("onTransferCompleted")
        if (diffTime == null) diffTime = (clock.nowMillis() - startTime).toInt()
    }

    abstract val dataSize: Int
    abstract val token: ByteArray?
    val speed: Long
        get() = if (diffTime == null) -1 else (dataSize.toDouble() / (diffTime!! / 1000.0)).toLong()
    val percentOfLoss: Double?
        get() = if (numberOfMessages == 0) null else numberOfResend.toDouble() * 100 / (numberOfResend + numberOfMessages)
    abstract val isIncoming: Boolean
    override fun onResend() {
        numberOfResend++
    }

    fun incrementNumberOfMessage() {
        numberOfMessages++
    }
}